java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CommonRequestHandler.java

```diff
@@ -0,0 +1,209 @@
+
+@NoArgsConstructor(onConstructor_ = @Inject)
+@Log
+public class CommonRequestHandler {
+
+    //we have ~540 total in Cloud Function connection, so can have generous values here
+    final int SOURCE_API_REQUEST_CONNECT_TIMEOUT = 30_000;
+    final int SOURCE_API_REQUEST_READ_TIMEOUT = 300_000;
+
+    @Inject ConfigService config;
+    @Inject RulesUtils rulesUtils;
+    @Inject SourceAuthStrategy sourceAuthStrategy;
+    @Inject ObjectMapper objectMapper;
+    @Inject SanitizerFactory sanitizerFactory;
+    @Inject Rules rules;
+
+    private Sanitizer sanitizer;
+
+    @SneakyThrows
+    public HttpEventResponse handle(HttpEventRequest request) {
+
+        //TODO: cache this?? no point in re-parsing config each time ...
+        this.sanitizer = sanitizerFactory.create(sanitizerFactory.buildOptions(config, rules));
+
+        boolean isHealthCheck =
+            request.getHeader(ControlHeader.HEALTH_CHECK.getHttpHeader()).isPresent();
+        if (isHealthCheck) {
+            return doHealthCheck(request);
+        }
+
+        HttpRequestFactory requestFactory = getRequestFactory(request);
+
+        // re-write host
+        //TODO: switch on method to support HEAD, etc
+        URL targetUrl = buildTarget(request);
+
+        boolean skipSanitizer = config.isDevelopment() &&
+            isRequestedToSkipSanitizer(request);
+
+        HttpEventResponse.HttpEventResponseBuilder builder = HttpEventResponse.builder();
+
+        if (skipSanitizer || sanitizer.isAllowed(targetUrl)) {
+            log.info("Proxy invoked with target " + URLUtils.relativeURL(targetUrl));
+        } else {
+            builder.statusCode(403);
+            log.warning("Attempt to call endpoint blocked by rules: " + targetUrl + "; rules " + objectMapper.writeValueAsString(rules.getAllowedEndpointRegexes()));
+            return builder.build();
+        }
+
+        com.google.api.client.http.HttpRequest sourceApiRequest;
+        try {
+            sourceApiRequest =
+                requestFactory.buildGetRequest(new GenericUrl(targetUrl.toString()));
+        } catch (IOException e) {
+            builder.statusCode(500);
+            builder.body("Failed to authorize request; review logs");
+            log.log(Level.WARNING, e.getMessage(), e);
+
+            //something like "Error getting access token for service account: 401 Unauthorized POST https://oauth2.googleapis.com/token,"
+            log.log(Level.WARNING, "Confirm oauth scopes set in config.yaml match those granted via Google Workspace Admin Console");
+            return builder.build();
+        }
+
+        //TODO: what headers to forward???
+
+        //setup request
+        sourceApiRequest
+            .setThrowExceptionOnExecuteError(false)
+            .setConnectTimeout(SOURCE_API_REQUEST_CONNECT_TIMEOUT)
+            .setReadTimeout(SOURCE_API_REQUEST_READ_TIMEOUT);
+
+        //q: add exception handlers for IOExceptions / HTTP error responses, so those retries
```

Comment:
No, we have the retry logic in Worklytics adapting to each case. IMHO Proxy should be _dumb_ in that regard.